node/src/components/linear_chain/state.rs

```diff
@@ -410,6 +413,91 @@ mod tests {
+        let block = Block::random(&mut rng);
+        let block_hash = *block.hash();
+        let block_era = block.header().era_id();
+        // Store some pending finality signatures
+        let _sig_a = add_pending(&mut lc, block_hash, block_era, true);
+        let _sig_b = add_pending(&mut lc, block_hash, block_era, false);
+        let _sig_c = add_pending(&mut lc, block_hash, block_era, false);
+
+        let execution_results = HashMap::new();
+        let outcomes = lc.handle_new_block(Box::new(block), execution_results);
+        // None of the signatures' creators have been confirmed to be bonded yet.
+        // We should not gossip/store/announce any signatures yet.
+        assert!(matches!(
+            &*outcomes,
+            [Outcome::StoreBlock(_), Outcome::StoreExecutionResults(_, _)]
+        ));
+    }
+
+    #[test]
+    fn new_block_bonded_pending_sigs() {
+        let mut rng = TestRng::new();
+        let protocol_version = ProtocolVersion::V1_0_0;
+        let mut lc = LinearChain::new(protocol_version, 1u64, 1u64);
+        let block = Block::random(&mut rng);
+        let block_hash = *block.hash();
+        let block_era = block.header().era_id();
+        // Store some pending finality signatures
+        let sig_a = add_pending(&mut lc, block_hash, block_era, true);
+        let sig_b = add_pending(&mut lc, block_hash, block_era, false);
+        let _sig_c = add_pending(&mut lc, block_hash, block_era, false);
+        // Mark two of the creators as bonded.
+        mark_bonded(&mut lc, sig_a.clone());
+        mark_bonded(&mut lc, sig_b.clone());
+        let execution_results = HashMap::new();
+        let outcomes = lc.handle_new_block(Box::new(block.clone()), execution_results.clone());
+        // `sig_a` and `sig_b` are valid and created by bonded validators.
+        let expected_outcomes = {
+            let mut tmp = vec![];
+            let mut block_signatures = BlockSignatures::new(block_hash, block_era);
+            block_signatures.insert_proof(sig_a.public_key.clone(), sig_a.signature);
+            block_signatures.insert_proof(sig_b.public_key.clone(), sig_b.signature);
+            tmp.push(Outcome::StoreBlockSignatures(block_signatures));
+            tmp.push(Outcome::Gossip(Box::new(sig_a.clone()))); // Only `sig_a` was created locally and we don't "regossip" incoming signatures.
+            tmp.push(Outcome::AnnounceSignature(Box::new(sig_a)));
+            tmp.push(Outcome::AnnounceSignature(Box::new(sig_b)));
+            tmp.push(Outcome::StoreBlock(Box::new(block)));
+            tmp.push(Outcome::StoreExecutionResults(
+                block_hash,
+                execution_results,
+            ));
+            tmp
+        };
+        // Verify that all outcomes are expected.
+        for outcome in outcomes {
+            assert!(
+                expected_outcomes.iter().any(|e| e == &outcome),
+                "unexpected outcome"
+            );
+        }
```

Comment:
That only asserts that `outcomes` is a subset of `expected_outcomes`. Maybe it would be best to collect both into a set, and then compare the sets.

Reply:
I knew it was too good to be true.

Reply:
I can't collect them into a set: it cannot be `BTreeSet` b/c I can't derive `PartialOrd, Ord` on the `Outcome` enum and it can't be a `HashSet` as I can't derive `Hash` on that enum either.

Reply:
Then you could just do the other check as well: That `expected_outcomes` is a subset of `outcomes`.

Reply:
Done in [`61a1ac0` (#1291)](https://github.com/CasperLabs/casper-node/pull/1291/commits/61a1ac07350ed98efaa052725463588407f7761d), please verify.